Support Configuring Arrow RecordBatch Writers via SQL Statement Options

devinjdangelo opened the pull request:
## Which issue does this PR close?

Closes #7322
Closes #7298

## Rationale for this change

Currently, the only way to configure how files are written out as a result of `INSERT` or `COPY` statements is via session level configuration. Additionally, string parsing of some special arbitrary SQL statement options are handled in various places in DataFusion on an adhoc basis (such as the current per_thread_output setting in `COPY `).

This PR aims to consolidate logic for parsing arbitrary SQL statement options and support reuse of code, including in downstream systems which may choose to support their own special arbitrary options. 

## What changes are included in this PR?

- Move existing parquet setting string parsing logic to datafusion-common
- Create abstractions to make working with tuples of arbitrary string options easier
- Create a new DataFusionError type for unsupported options
- Implement string parsing logic for CSV and JSON settings
- Support writing compressed CSV and JSON files (this one is a DataFusion specific write option)
- Rename `per_thread_output` option to `single_file_output` as I believe this is less ambiguous in DataFusion (DuckDB uses the `per_thread_output` option, but the reference to threads doesn't really make sense in DataFusion). This also makes more sense in a table context where a listing table can be backed by a single file or not.

An example of a new query supported by this PR:

```sql
COPY source_table TO 
'./table' 
(format parquet, 
single_file_output false, 
compression 'zstd(10)');
```

### Notable work that is important for follow ons
- Support these options in the DataFrame::write_* functions without requiring passing of strings (current thinking is to allow passing a fully formed writer builder object directly).
- Improve test coverage of various combinations of write options.

## Are these changes tested?

Yes, via existing tests and some new ones. However, there is now a very large number of possible option combinations. We may want to implement some dynamic test case generation framework to ensure no specific combination of options will fail.

## Are there any user-facing changes?

Yes, writes now support most options that the various arrow writers support via SQL statements. The primary exception right now is column specific parquet options (i.e. setting an encoding for only 1 column). Additional work will be need to support that.

devinjdangelo commented on datafusion/common/src/error.rs:
```diff
@@ -73,6 +73,9 @@ pub enum DataFusionError {
     Plan(String),
+    /// This error happens when an invalid or unsupported option is passed
+    /// in a SQL statement
+    InvalidOption(String),
```
If this causes any issues with the ongoing work around DataFusionError, I'm happy to roll it back. I created this because I didn't feel any existing error type seemed to make sense for an invalid option.

devinjdangelo commented on datafusion/common/src/file_options/mod.rs:
```diff
@@ -0,0 +1,269 @@
+/// around a "writer builder" type (e.g. arrow::csv::WriterBuilder)
+/// plus any DataFusion specific writing options (e.g. CSV compression)
+#[derive(Clone, Debug)]
+pub enum FileTypeWriterOptions {
```
This enum was proposed by @alamb in #6569 and is the key new abstraction in this PR.

devinjdangelo commented on datafusion/core/src/datasource/listing/table.rs:
```diff
@@ -1828,7 +1862,7 @@ mod tests {
         .await
         .expect_err("Example should fail!");
-        assert_eq!("Error during planning: zstd compression requires specifying a level such as zstd(4)", format!("{e}"));
+        assert_eq!("Invalid Option: zstd compression requires specifying a level such as zstd(4)", format!("{e}"));
```
This example I think shows why I wanted to create a new error type. "Error during planning" I think would be more confusing to an end user than just a straightforward "Invalid Option" error.

devinjdangelo commented on datafusion/sqllogictest/test_files/copy.slt:
```diff
@@ -139,20 +160,11 @@ select * from validate_single_json;
+
 # Copy from table with options
-query IT
+query error DataFusion error: Invalid Option: Found unsupported option row_group_size with value 55 for JSON format!
```
As discussed in #6569, this query now throws an error since row_group_size does not make sense in the context of writing JSON files.

alamb commented:
> Yes, via existing tests and some new ones. However, there is now a very large number of possible option combinations. We may want to implement some dynamic test case generation framework to ensure no specific combination of options will fail.



FWIW I don't think it is necessary to test all the combinations -- that is largely tested upstream in the arrow writer so testing in DataFusion that all the plumbing is hooked up correctly is likely the most important thing

alamb commented on datafusion/common/src/file_options/csv_writer.rs:
```diff
@@ -0,0 +1,91 @@
+    /// However, WriterBuilder does not allow public read access to the
+    /// has_header parameter.
+    pub has_header: bool,
+    // TODO: expose a way to read has_header in arrow create
```
I filed https://github.com/apache/arrow-rs/issues/4735

```suggestion
    // TODO: expose a way to read has_header in arrow create
    // https://github.com/apache/arrow-rs/issues/4735
```

alamb commented on datafusion/common/src/file_options/mod.rs:
```diff
@@ -0,0 +1,269 @@
+        match self {
+            FileTypeWriterOptions::Parquet(opt) => Ok(opt),
+            _ => Err(DataFusionError::Internal(
+                "Expected parquet options but found options for a different FileType!"
```
I think it would help usablity if these errors contained the name of the other type. Also should these be `DataFusionError::InvalidOption`? `DataFusionError::Internal` is supposed to signal a bug

```suggestion
                "Expected parquet options but found options for: {}", self.name()
```

The same comment applies to the other `try_into*` functions as well

devinjdangelo replied:
I chose internal error here because this error is only hit if somehow the calling code expects options for format A but actually options for format B were set. That should not be possible based on any input a user provides. It should only happen if we have an internal logic bug (like calling try_into_json within parquet_sink).

InvalidOption error happens if the user provides invalid options, but invalid options alone cannot lead to this particular error (at least I can't think of a way without some internal bug).

alamb commented on datafusion/common/src/file_options/parquet_writer.rs:
```diff
@@ -0,0 +1,163 @@
+/// Constructs a default Parquet WriterPropertiesBuilder using
+/// Session level ConfigOptions to initialize settings
+fn default_builder(options: &ConfigOptions) -> Result<WriterPropertiesBuilder> {
+    let parquet_context = &options.execution.parquet;
```
nit: calling this `parquet_options` might be clearer

alamb commented on datafusion/common/src/file_options/parse_utils.rs:
```diff
@@ -0,0 +1,183 @@
+}
+
+/// Parses datafusion.execution.parquet.encoding String to a parquet::basic::Encoding
+pub(crate) fn parse_encoding_string(
```
It would be great to move this stuff upstream into a `FromStr` implementation in the parquet crate. I can  perhaps file a ticket to do so.

devinjdangelo replied:
I believe you already did! https://github.com/apache/arrow-rs/issues/4693

alamb commented on datafusion/core/src/datasource/physical_plan/csv.rs:
```diff
@@ -1191,6 +1191,8 @@ mod tests {
             Field::new("c2", DataType::UInt64, false),
         ]));
 
+        println!("{out_dir}");
```
I think this might be left over? Since it is part of a test I think it is fine to leave in, just thought I would point it out

alamb commented on datafusion/core/src/datasource/physical_plan/mod.rs:
```diff
@@ -94,12 +93,14 @@ pub struct FileSinkConfig {
     /// regardless of input partitioning. Otherwise, each table path is assumed to be a directory
     /// to which each output partition is written to its own output file.
-    pub per_thread_output: bool,
+    pub single_file_output: bool,
```
I agree the new name is much nicer

alamb reviewed (commented):
This is great @devinjdangelo  -- thank you! I also tried it out locally and it worked great. It is not that often that I review 1500+ line PRs and have so few comments 🥇 

The only thing I think I think this PR needs prior to merging is test coverage for config parsing (as you mentioned in your description)

I also think we should document the available options (but that can be done as a follow on PR)

Local test

```shell
~/Software/target-df2/release/datafusion-cli -c "copy 'traces_nd.parquet' to 'out' (format parquet, single_file_output false, compression 'zstd(10)')"
DataFusion CLI v30.0.0
+---------+
| count   |
+---------+
| 5185717 |
+---------+
1 row in set. Query took 23.679 seconds.
```

alamb commented:
I also took the liberty of merging this branch to `main` to resolve a conflict (I am so excited to see this feature almost ready!)

devinjdangelo commented:
Thank you for the review @alamb! All of your suggestions make sense. I plan to get to updating this pr and adding some additional test cases within the next few days.

I also noticed your local test took 23s to execute. I am hoping to (finally) get to improving that performance soon!

alamb commented:
> I also noticed your local test took 23s to execute. I am hoping to (finally) get to improving that performance soon!

thank **you** for all your help!

alamb commented:
Let me know if you need help finishing up this PR @devinjdangelo  -- I am on vacation this week so I can focus my energy and reviews on things I think are the most interesting -- and this is one of them!

devinjdangelo commented:
Thanks @alamb! I just pushed up some updates addressing review comments and adding a few more tests. The main thing remaining is unit tests which set options via a `StatementOptions` struct and inspect the resulting `FileTypeWriterOptions` to ensure the settings actually were set as expected. The current sqllogictests are only ensuring that passing the option does not result in an error, not that the setting actually had an effect on the underlying `RecordBatchWriter`.

I expect to get back to work on unit tests later today (after 5pm EST).  Feel free to take a crack at it before I get to it (or just enjoy your vacation)!

Also noticed many clippy warnings locally unrelated to the changes in this PR. I believe that is related to the clippy updates upstream, so may just need to merge main again.

devinjdangelo commented:
I pushed up some unit tests. The CSV option parsing unit test cannot validate options were set properly on `csv::WriterBuilder` since actually none of its properties are externally readable. We should update this unit test when https://github.com/apache/arrow-rs/issues/4735 is closed out.

Also, solving my local clippy errors was as simple as running `rustup update`.

alamb reviewed (approved):
Thank you @devinjdangelo  -- this looks epic! Really nice work
